crypto/src/hasher/mod.rs

```diff
@@ -0,0 +1,278 @@
+//!
+//! An abstract hashing interface and associated traits
+//!
+//! **Example**
+//! ```rust
+//! use mina_crypto::hasher::{create_legacy, Hashable, Hasher, ROInput};
+//!
+//! #[derive(Clone)]
+//! struct Example {
+//!     a: u32,
+//!     b: u64,
+//! }
+//!
+//! impl Hashable for Example {
+//!     type D = ();
+//!
+//!     fn to_roinput(self) -> ROInput {
+//!         let mut roi = ROInput::new();
+//!         roi.append_u32(self.a);
+//!         roi.append_u64(self.b);
+//!         roi
+//!     }
+//!
+//!     fn domain_string(_: Option<Self>, _: &Self::D) -> String {
+//!         format!("Example")
+//!     }
+//! }
+//!
+//! // Usage example
+//! let mut hasher = create_legacy::<Example>(());
+//! let out = hasher.hash(Example {a: 1, b: 2});
+
+pub mod poseidon;
+pub mod roinput;
+
+use ark_ff::PrimeField;
+use mina_curves::pasta::Fp;
+use o1_utils::FieldHelpers;
+pub use roinput::ROInput;
+
+/// The domain parameter trait is used during hashing to convey extra
+/// arguments to domain string generation.  It is also used by generic signing code.
+pub trait DomainParameter: Clone {
+    /// Conversion into vector of bytes
+    fn into_bytes(self) -> Vec<u8>;
+}
+
+impl DomainParameter for () {
+    fn into_bytes(self) -> Vec<u8> {
+        vec![]
+    }
+}
+
+impl DomainParameter for u32 {
+    fn into_bytes(self) -> Vec<u8> {
+        self.to_le_bytes().to_vec()
+    }
+}
+
+impl DomainParameter for u64 {
+    fn into_bytes(self) -> Vec<u8> {
+        self.to_le_bytes().to_vec()
+    }
+}
+
+/// Interface for hashable objects
+///
+/// Mina uses fixed-length hashing with domain separation for each type of object hashed.
+/// The prior means that `Hashable` only supports types whose size is not variable.
+///
+/// **Important:** The developer MUST assure that all domain strings used throughout the
+/// system are unique and that all structures hashed are of fixed size.
+///
+/// Here is an example of how to implement the `Hashable` trait for am `Example` type.
+///
+/// ```rust
+/// use mina_crypto::{
+///     hasher::{Hashable, ROInput},
+///     signer::NetworkId,
+/// };
+///
+/// #[derive(Clone)]
+/// struct Example;
+///
+/// impl Hashable for Example {
+///     type D = NetworkId;
+///
+///     fn to_roinput(self) -> ROInput {
+///         let roi = ROInput::new();
+///         // Serialize example members
+///         // ...
+///         roi
+///     }
+///
+///     fn domain_string(_: Option<Self>, network_id: &NetworkId) -> String {
+///        match network_id {
+///            NetworkId::MAINNET => "ExampleMainnet",
+///            NetworkId::TESTNET => "ExampleTestnet",
+///        }.to_string()
+///    }
+/// }
+/// ```
+///
+/// See example in [`ROInput`] documentation
+pub trait Hashable: Clone {
+    /// Generic domain string argument type
+    type D: DomainParameter;
+
+    /// Serialization to random oracle input
+    fn to_roinput(self) -> ROInput;
+
+    /// Generate unique domain string of length `<= 20`.
```

Comment:
perhaps clarify in the doc that the uniqueness and length bound are not imposed for free, but should be enforced by the developer implementing the trait

Reply:
can you assert this somewhere though?

Reply:
Yes and it already is. 👍🏼

Reply:
There's already a comment on the trait warning that the uniqueness bound must be assured by the developer, but I will also add a note about this on the method itself.  Good suggestion - thanks.